Tests/swift-sdk-swift-tests/IterableAutoRegistrationTests.swift

```diff
@@ -0,0 +1,207 @@
+            networkSession.callback = {(_, _, _)in
+                // Second call, API call to disable endpoint
+                expectation3.fulfill()
+                TestUtils.validate(request: networkSession.request!, requestType: .post, apiEndPoint: ITBConsts.apiEndpoint, path: ENDPOINT_DISABLE_DEVICE, queryParams: [(name: ITBL_KEY_API_KEY, value: IterableAutoRegistrationTests.apiKey)])
+                let body = networkSession.getRequestBody() as! [String : Any]
+                TestUtils.validateElementPresent(withName: ITBL_KEY_TOKEN, andValue: (token as NSData).iteHexadecimalString(), inDictionary: body)
+                TestUtils.validateElementPresent(withName: ITBL_KEY_EMAIL, andValue: "user1@example.com", inDictionary: body)
+            }
+            
+            IterableAPI.email = "user2@example.com"
+        }
+        IterableAPI.register(token: token)
+        
+        // only wait for small time, supposed to error out
+        wait(for: [expectation1, expectation2, expectation3], timeout: testExpectationTimeout)
+    }
+    
+    func testDoNotCallDisableAndEnableWhenSameValue() {
+        let expectation = XCTestExpectation(description: "testDoNotCallDisableAndEnableWhenSameValue")
+        
+        let networkSession = MockNetworkSession(statusCode: 200)
+        let config = IterableConfig()
+        config.pushIntegrationName = "my-push-integration"
+        config.autoPushRegistration = true
+        // notifications are enabled
+        let notificationStateProvider = MockNotificationStateProvider(enabled: true) // Notifications are on.
+        
+        IterableAPI.initialize(apiKey: IterableAutoRegistrationTests.apiKey, config:config, networkSession: networkSession, notificationStateProvider: notificationStateProvider)
+        let email = "user1@example.com"
+        var registerForRemoteNotificationsCount = 0
+        notificationStateProvider.callback = {
+            registerForRemoteNotificationsCount += 1
+        }
+        IterableAPI.email = email
+        let token = "zeeToken".data(using: .utf8)!
+        networkSession.callback = {(_, _, _) in
+            // first call back will be called on register
+            TestUtils.validate(request: networkSession.request!, requestType: .post, apiEndPoint: ITBConsts.apiEndpoint, path: ENDPOINT_REGISTER_DEVICE_TOKEN, queryParams: [(name: ITBL_KEY_API_KEY, value: IterableAutoRegistrationTests.apiKey)])
+            networkSession.callback = {(_, _, _)in
+                // Second callback should not happen
+                XCTFail("Should not call disable")
+            }
+            // set same value
+            IterableAPI.email = email
+            
+            // wait for 1 second to make sure that registerForRemoteNotifications is not called
+            DispatchQueue.main.asyncAfter(deadline: .now() + 1, execute: {
+                XCTAssertEqual(registerForRemoteNotificationsCount, 1) // should be called only once (when setting user1@example.com).
```

Comment:
What do you think about modifying the mocks to take expectations and then fulfill them in the callback? Like, add an expectation parameter to MockNotificationStateProvider, and then you'd be able to use `expectedFulfillmentCount` to verify the call count.

Reply:
Brilliant!